Better UX for IV drip

MTandi opened the pull request:
## About The Pull Request

Resolves #74752 with the following:

IV drip screntips now mention that you can toggle it between Min and Max transfer rate without opening UI with Alt+Click

Added buttons to min or max out the transfer rate within the UI, without the need to use slider and made this control available regardless of container and object.

<img width="300" alt="XcoPcE6Su4" src="https://user-images.githubusercontent.com/3625094/233209370-d95f32c6-6044-43b7-8379-b6130d561d32.png">

Also fixed animation states.

## Why It's Good For The Game

You couldn't start injection with the desired flow rate before this. 

## Changelog

:cl:
qol: IV drip flow rate can be changed without the container or object attached
fix: IV drip animation states fixed
/:cl:

MTandi committed:
better ux for iv drip

msgerbs commented on tgui/packages/tgui/interfaces/IVDrip.tsx:
```diff
@@ -133,7 +133,7 @@ export const IVDrip = (props, context) => {
             ) : (
               <LabeledList.Item label="Object">
                 <Tooltip content="Drag the cursor from the drip and drop it on an object to connect.">
-                  <NoticeBox my={0.7}>No object hasObjectAttached.</NoticeBox>
```
lol...

Mothblocks reviewed (changes requested):
How on earth is this better UX than the PR that changed the default?

I'm the one who merged it and I'm very aware of alt-click. The original PR is still correct that any reasonable person would expect attaching an IV drip to start also performing the IV drip action. Shoving more text into the examine is not at all an improvement over the PR you're reverting. Do not do that.

Instead:
- Keep your tgui changes
- Change the `desc_controls` to a contextual screentip
- Keep the new transfer rate

msgerbs commented:
Seems like the only thing that needs to be changed to address the actual UX issue is to be able to set the inject speed when you insert the container, rather than having to set it after you attach it. That makes less sense from a real-world perspective of how IVs work but it makes more sense from a game UX standpoint where you really don't want to just inject a random chemical at full speed immediately.

Mothblocks commented:
How often do you want less than the maximum output? If the answer is extremely common, then that should probably be the default. Real solution to this is probably sprites that help you know roughly how much you're going to send before you configure it in the UI, but as a medical player most of the time I really don't care

msgerbs commented:
> How often do you want less than the maximum output? If the answer is extremely common, then that should probably be the default. Real solution to this is probably sprites that help you know roughly how much you're going to send before you configure it in the UI, but as a medical player most of the time I really don't care and just alt-click right away

The original complaint is that you *can't* set the amount until you've already connected. It starts at max flow, so if you want to drip, say, 1u instead of 5u, you have to connect it up and then change it, by which time it will have probably already given several 5u doses.

Mothblocks commented:
Yeah, and the tgui changes to fix that are fine--what I'm arguing against is that reverting the PR is better UX than keeping it

msgerbs commented:
I agree with that. Sorry, it seemed like you were disagreeing with me.

MTandi commented:
> Instead:
> 
> * Keep your tgui changes
> * Change the `desc_controls` to a contextual screentip
> * Keep the new transfer rate

Sorry, I didn't follow your arguments here. The point is that the new transfer rate brings issues in non-medbay cases as stated in #74752. And this set of bullet points won't resolve the issue, the previous PR brings. I assume that by "new transfer rate" you mean the MAX rate that is currently on master, right?

>How often do you want less than the maximum output? If the answer is extremely common, then that should probably be the default.

Yes, that is why changed the default rate back to 0.

So in this PR you suggest to:

* Change the default transfer rate back to maximum and keep the #74752 open
* Change the desc_controls to a contextual screentip

Right?

MTandi commented:
> Seems like the only thing that needs to be changed to address the actual UX issue is to be able to set the inject speed when you insert the container, rather than having to set it after you attach it. That makes less sense from a real-world perspective of how IVs work but it makes more sense from a game UX standpoint where you really don't want to just inject a random chemical at full speed immediately.

I would make it drip chems onto the floor when you attach container, set non-zero transfer rate but don't have an object attached.

Imagine clown dragging around an IV drip that is leaving a space lube trail behind 😄

Mothblocks commented:
> The point is that the new transfer rate brings issues in non-medbay cases as stated in https://github.com/tgstation/tgstation/issues/74752. And this set of bullet points won't resolve the issue, the previous PR brings

I don't understand why it doesn't. Your tgui changes make it easy to change the throughput without turning on the IV drip. Going out of your way to do that is fine because it is very much the edge case--you imply it yourself when you say "non-medbay cases". I think the medbay cases for the IV drips should take priority. If there are jobs that actively use IV drips at lower frequencies, then we should give them a specialized IV drip that has a lower max threshold and some distinct style (e.g. like a hand-me-down or something).

> > How often do you want less than the maximum output? If the answer is extremely common, then that should probably be the default.
> Yes, that is why I changed the default rate back to 0.

Yes to what? I barely ever worry about the throughput rate for medbay--I attach it and turn it on. Being able to skip the "turn it on" step is a UX improvement to people who are less familiar with the IV drips than me.

> So in this PR you suggest to:
> Change the default transfer rate back to maximum and keep the https://github.com/tgstation/tgstation/issues/74752 open
> Change the desc_controls to a contextual screentip
> Right?

You have my changes correct, but I don't see why this doesn't close #74752. You are adding the ability to use the IV drip without immediately giving the max output when you add the tgui controls, which you can presumably use before attaching it.

I feel like there is a miscommunication going on somewhere. Can you help me figure that out?

msgerbs commented:
I think the confusion is that they didn't actually make any tgui changes to fix the issue, they just added an on/off button to the existing UI. You still have to connect it first before the slider or button appear.

msgerbs commented:
> Okay, yes that is the miscommunication. So then I'm not sure that #74752 needs to remain open with or without this PR if the core of it, that you can't start injection rates with small amounts, is incorrect.

The issue should still stay open. You can't start injection with anything other than 5u, because you have to connect the IV, at which point it will start injecting at 5u/t, then you can open the UI and change the amount. The fix is to not hide the injection rate slider when disconnected.

MTandi commented:
@msgerbs is right, my changes are:

- Added a UI button that basically does what Alt+Click already did.
- Added a tooltip because it seems like the author of the original PR wasn't aware of Alt+Click.
- Reverted the default rate for 5 u/sec back to 0 u/sec, so you, unfortunately **HAVE** to alt+click or tap the UI button for default use case, but people are now either aware of the right click or have to only one click in UI instead of dragging the slider.

I did this because the original PR broke many cases to save one click for the blood injection by making IV drip transferring 5 u/sec by default, so it shouldn't have been merged in that shape. This is what #74752 is about.

So if I change the rate from 0 u/sec. back to 5 u/sec., the animation will stay broken and use cases when you need small amounts of chems injected will also stay broken. 

And this is not just about non-medbay stuff. Imagine an IV drip with Omnizine. It's a chem that you don't need much to heal a person and it overdoses at 30 units, dealing damage. @Mothblocks in case of Omnizine, you believe it is correct that it will also be automatically injected at 5 u/sec rate by default, despite the risk of wasting chemicals or causing overdose?

MTandi commented:
> The issue should still stay open. You can't start injection with anything other than 5u, because you have to connect the IV, at which point it will start injecting at 5u/t, then you can open the UI and change the amount. The fix is to not hide the injection rate slider when disconnected.

So you suggest to always pre-configure the transfer rate so when you attach it, it will always automatically start draining or injecting at the set rate?

MTandi committed:
change rate wihout container

MTandi committed:
rmd to screentip and fixes

msgerbs commented:
Yes, exactly like that. My only suggestion would be a way to enter the exact amount you want to transfer, other than that I think this is perfect.

MTandi commented:
> Yes, exactly like that. My only suggestion would be a way to enter the exact amount you want to transfer, other than that I think this is perfect.

The slider component already allows to enter exact amount if you double-click the slider.

msgerbs commented:
Wait seriously?! Well, TIL. I think this is good then, hopefully Mothblocks agrees.

Mothblocks commented:
> And this is not just about non-medbay stuff. Imagine an IV drip with Omnizine. It's a chem that you don't need much to heal a person and it overdoses at 30 units, dealing damage. @Mothblocks in case of Omnizine, you believe it is correct that it will also be automatically injected at 5 u/sec rate by default, despite the risk of wasting chemicals or causing overdose?

Yes, because you are still describing an edge case. It is an edge case we support, but it is not nearly as widespread as the expectation you will be putting blood inside it to be worth adjusting the default for.

> So if I change the rate from 0 u/sec. back to 5 u/sec., the animation will stay broken and use cases when you need small amounts of chems injected will also stay broken.

Then the issue report should be isolated to just this, rather than that it can't be adjusted before using, which does not appear to be true.

msgerbs commented:
> > So if I change the rate from 0 u/sec. back to 5 u/sec., the animation will stay broken and use cases when you need small amounts of chems injected will also stay broken.
> 
> Then the issue report should be isolated to just this, rather than that it can't be adjusted before using, which does not appear to be true.

I agree the animation issue should be a separate issue, but this is indeed true. Without this PR it is not possible to adjust before you attach it. They just updated the PR to add that capability.

MTandi commented:
@Mothblocks all done, but the Alt + Left-Click hint remained in the examine description, because it doesn't have such parameter. I moved the RMB to screentip though.

![image](https://user-images.githubusercontent.com/3625094/233210522-10e30aa1-6ac8-449d-902e-fa7b3c4fcfd9.png)

Mothblocks commented on code/game/machinery/iv_drip.dm:
```diff
@@ -56,6 +59,10 @@
 	interaction_flags_machine |= INTERACT_MACHINE_OFFLINE
+	AddElement( \
+		/datum/element/contextual_screentip_bare_hands, \
+		rmb_text = "Detach / Eject / Configure", \
```
Surely it doesn't do all at the same time, every time?

Mothblocks replied:
If what RMB does is contextual (it definitely is, can't eject something you don't have), then you can't use this element.

You instead want to override `add_context`, and call `register_context` in Initialize. Ctrl+shift+F for `/add_context` to see examples.

Mothblocks replied:
You want the RMB text to change depending on what it's going to do.

MTandi committed:
screentips

MTandi committed:
rmb detach doesnt open ui
